### DISCUSSION TITLE
[Avram] What is the correct way to retrieve one record from db, and invoke methods on it.

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1700] [Q&A] [answered]
We assume users table, exists a `open_id` column can be retreive the unique record, but it can be nil if `open_id` is not correct.

### Solution 1: works

```cr
open_id = "open_id"
user : User = UserQuery.new.open_id(open_id).first # this may raise exception.

user.account
user.password_digest
```

It works, but, first maybe raise exception instead if no matched record, so we should do rescue, right?

### Solution 2 works

```cr
open_id = "open_id"
user = UserQuery.new.open_id(open_id).first? # it should be type User | Nil

return nil if user.blank?

# return nil if user.locked_at  # this line code not work, because `undefined method 'locked_at' for Nil`

user.try &.locked_at
```

But, all method invoke on user must use &, it so tediously.

### Solution 3 not work

```cr
user : User | Nil

user_query = UserQuery.new.open_id(open_id)

if user_query.any?
  user  = user_query.first
end
```

here, user type is `User | Nil`,  i have user_query.any? if condition, why still this?

So, my question is, what is the correct way to do those things in varam? i am still newbie to crystal and lucky, if there is a another recommended why to do this?

**Comment by jwoertink** [comment, marked as answer]
In Crystal, you can assign the nilable value to a variable within the `if` condition, and the compiler will know the type.

```crystal
if user = UserQuery.new.open_id("open_id").first?
  # this works because Crystal compiler knows `user` exists here.
  user.account
end
```

You can also do something like

```crystal
if user_query.any?
  user = user_query.first.as(User)
end
```

The `.as(User)` will tell the compiler that the object is that type if a Union is possible (`User | Nil`).

**Reply by zw963** [reply]
Thanks for reply, i consider the as solution is better.

 > In Crystal, you can assign the nilable value to a variable within the if condition, and the compiler will know the type.

 For the first if solution, But it not work in following case:

```cr
def some_method
  open_id = "open_id"
  user = UserQuery.new.open_id(open_id).first? # it should be type User | Nil

  return nil if user.blank?    # this if/return guard check is not valid.

  return nil if user.locked_at  #  raise Error `undefined method 'locked_at' for Nil`
end  
```

Why?

**Reply by jwoertink** [reply]
I think that's related to how Crystal does the nil guard checks. It doesn't work when using one-liner, only with the `if/end` version.

```crystal
def some_method
  open_id = "open_id"
  if user = UserQuery.new.open_id(open_id).first?
    return nil if user.locked_at
    # do other things with user
  end
end
```

**Reply by zw963** [reply]
So, any cases, we have to wrap all following `use user object code` into a if block, to make compiler pass check, right?

**Reply by jwoertink** [reply]
Also, if it helps any, you can watch this video https://luckycasts.com/videos/handling-nil Maybe it will explain better?

**Reply by jwoertink** [reply]
> So, any cases, we have to wrap all following `use user object code` into a if block, to make compiler pass check, right?

Yes. If the object is potentially nilable, then you have to wrap in to if block. :+1:

**Reply by grepsedawk** [reply]
This isn't happening because of the guard clause, this is happening because the `user`, in this context, can change between the guard clause and the _actual_ line of code.
It appears like `user` might be a getter for a @user ivar?
This can be handled by assigning it to a variable scoped to the method, like @jwoertink expressed, my favorite for this pattern is

```crystal
UserQuery.new.open_id(open_id).first?.try do |user|
  return nil if user.locked_at?
  # do other things with user, safely
end
```

The try(&) is comparable to ruby's then, and provides a clear block of where "user" is in a nil-safe scope :)

Happy Hacking!

**Reply by robacarp** [reply]
This is not true. I use early-exit return style code all the time. With this code:

```crystal
def try_it(var : Int32?)
  p typeof(var)
  return if var.nil?
  p typeof(var)
end


[1, 2, nil].each do |e|
  try_it e
  puts "==============="
end
```

You get this output:

```
(Int32 | Nil)
Int32
===============
(Int32 | Nil)
Int32
===============
(Int32 | Nil)
===============
```

After a `return if user.nil?` line, the rest of the lines in the method can operate knowing that `user` is not a nil.

See https://carc.in/#/r/dc10.

Edit: I see the problem in your sample.

For your code sample, try this:

```crystal
def some_method
  open_id = "open_id"
  user = UserQuery.new.open_id(open_id).first? # it should be type User | Nil

  return if user.nil?    # this if/return guard check is not valid.

  return nil if user.locked_at  #  raise Error `undefined method 'locked_at' for Nil`
end 
```

There is a difference between [#nil?](https://crystal-lang.org/api/1.4.1/Object.html#nil?:Bool-instance-method) (which is a compiler psuedo-method) and [#blank?](https://crystal-lang.org/api/1.4.1/String.html#blank?:Bool-instance-method) which is a method declared on String. #nil? informs the compiler and influences the code after it, while #blank? doesn't.

**Reply by grepsedawk** [reply]
@robacarp is correct, though there are times where an early return does NOT work, such as when you are calling a method/property related to an ivar that can be modified by another scope, and you must use a try pattern/if pattern

**Reply by Blacksmoke16** [reply]
```cr
def some_method
  open_id = "open_id"

  return unless (user = UserQuery.new.open_id(open_id).first?)

  user # : User
end
```

**Reply by robacarp** [reply]
Yes, crystal doesn't trust `@var`s types against guards. You can easily assign it to a local:

```crystal
class Foo
  getter bar : String?
  def initialize
  end

  def print_bar
    if @bar.nil?
      # @bar is still String?
    else
      # @bar is still String?
    end

    bar = @bar
    if bar.nil?
      # bar is known to be a Nil
    else
      # bar is known to be a String
    end
  end
end

```

**Reply by zw963** [reply]
@grepsedawk 

> It appears like user might be a getter for a @user ivar?

no, this is a variable, if it is a getter, sure this usage is invalid.

https://crystal-lang.org/reference/1.4/syntax_and_semantics/if_var.html#method-calls

**Reply by zw963** [reply]
> There is a difference between [#nil?](https://crystal-lang.org/api/1.4.1/Object.html#nil?:Bool-instance-method) (which is a compiler psuedo-method) and [#blank?](https://crystal-lang.org/api/1.4.1/String.html#blank?:Bool-instance-method) which is a method declared on String. #nil? informs the compiler and influences the code after it, while #blank? doesn't.

Cool, you point out the key!!

Yes, my code works after change `blank?` to `nil?`, you are awesome!

**Reply by zw963** [reply]
> ```crystal
>   return unless (user = UserQuery.new.open_id(open_id).first?)
> ```

This code works too, thank you, it equivalent to:

```crystal
user = UserQuery.new.open_id("open_id").first?

    return nil unless user

    return nil if user.locked_at  # work!
```

**Comment by zw963** [comment]
https://github.com/luckyframework/lucky/discussions/1700#discussioncomment-2987092 is consider as the expected answer.

Thanks all!
